File: v1/LLD/Design-Patterns/FactoryMethod/code/factory_method.py
"""
Factory Method Pattern — notification system example.
Run: python factory_method.py
"""

from abc import ABC, abstractmethod


class Notification(ABC):
    @abstractmethod
    def notify_user(self):
        ...


class EmailNotification(Notification):
    def notify_user(self):
        print("Sending an EMAIL notification.")


class SMSNotification(Notification):
    def notify_user(self):
        print("Sending an SMS notification.")


class NotificationFactory(ABC):
    @abstractmethod
    def create_notification(self) -> Notification:
        ...

    def send(self):
        notification = self.create_notification()
        notification.notify_user()


class EmailNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return EmailNotification()


class SMSNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return SMSNotification()


if __name__ == "__main__":
    email_factory = EmailNotificationFactory()
    sms_factory = SMSNotificationFactory()

    email_factory.send()
    sms_factory.send()

File: v1/LLD/Design-Patterns/FactoryMethod/code/FactoryMethod.cpp
// Factory Method Pattern — notification system example.
// Compile: g++ -std=c++14 FactoryMethod.cpp -o factory_method
// Run:     ./factory_method

#include <iostream>
#include <memory>

class Notification {
public:
    virtual void notifyUser() = 0;
    virtual ~Notification() = default;
};

class EmailNotification : public Notification {
public:
    void notifyUser() override {
        std::cout << "Sending an EMAIL notification." << std::endl;
    }
};

class SMSNotification : public Notification {
public:
    void notifyUser() override {
        std::cout << "Sending an SMS notification." << std::endl;
    }
};

class NotificationFactory {
public:
    virtual std::unique_ptr<Notification> createNotification() = 0;
    virtual ~NotificationFactory() = default;

    void send() {
        auto notification = createNotification();
        notification->notifyUser();
    }
};

class EmailNotificationFactory : public NotificationFactory {
public:
    std::unique_ptr<Notification> createNotification() override {
        return std::make_unique<EmailNotification>();
    }
};

class SMSNotificationFactory : public NotificationFactory {
public:
    std::unique_ptr<Notification> createNotification() override {
        return std::make_unique<SMSNotification>();
    }
};

int main() {
    EmailNotificationFactory emailFactory;
    SMSNotificationFactory smsFactory;

    emailFactory.send();
    smsFactory.send();

    return 0;
}
